Pioneer DDJ-400: Cycle through focused effects

etileb opened the pull request:
New FX section.

etileb committed:
Update Pioneer-DDJ-400-script.js

New FX section.

uklotzde commented on res/controllers/Pioneer-DDJ-400-script.js:
```diff
@@ -265,10 +265,14 @@ PioneerDDJ400.beatFxRightPressed = function(_channel, _control, value) {
-
-    engine.setValue("[EffectRack1_EffectUnit1]", "focused_effect", 3);
+    var focusfx = engine.getValue("[EffectRack1_EffectUnit1]", "focused_effect")+1;
+    if (engine.getValue("[EffectRack1_EffectUnit1]", "focused_effect")===3){
```
Please add a short comment what this code does. I guess cycling through the 2 focused effects 1 and 2?

Please also check the coding style, i.e. add spaces and indentation.

etileb replied:
I commented everything now, my first contrib on GitHub, sorry.

uklotzde reviewed (changes requested):
Please rebase on the 2.3 branch. This should also fix some pre-commit issues.

We prefer to merge fixes and improvements of mappings into the current release version and release them with the next minor version.

uklotzde commented:
As a *First-time contributor* please also sign our [Contributor License Agreement](https://docs.google.com/forms/d/e/1FAIpQLScC9QG327sjLL0eWftmfGUasxFFLxg6LCXJ2xHDYRzFIRqyiw/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ) which allows us to distribute your code in the *Apple App Store* (in case it is ever needed).

uklotzde commented:
Naming your PR like "DDJ-400: Cycle through focused effects" would help to guide reviewers.

The same applies to commit messages. Describe *what* has changed.

uklotzde commented on res/controllers/Pioneer-DDJ-400-script.js:
```diff
@@ -265,10 +265,14 @@ PioneerDDJ400.beatFxRightPressed = function(_channel, _control, value) {
-
-    engine.setValue("[EffectRack1_EffectUnit1]", "focused_effect", 3);
+    var focusfx = engine.getValue("[EffectRack1_EffectUnit1]", "focused_effect")+1;
+    if (engine.getValue("[EffectRack1_EffectUnit1]", "focused_effect")===3){
```
Ok, I see, the code is confusing, there are 3 effects.

- engine.getValue() should only be called once and assigned to variable, e.g. focused_effect
- if focused_effect >== 3 { focused_effect = 1 } else { focused_effect = focused_effect + 1 }

Be-ing commented:
What is the motivation for this change?

uklotzde commented:
The manual also needs to be updated if the changes in this PR should be merged:

https://manual.mixxx.org/2.3/ca/hardware/controllers/pioneer_ddj_400.html#effect-section-p-9

I agree with @Be-ing that we need to know more about the motivation for making an informed decision how to proceed. We don't own this controller and cannot try it out ourselves.

uklotzde commented:
The proposed changes would modify the focus behavior inconsistently, i.e. effect 1 and 2 could still be focused by dedicated buttons while effect 3 could only be reached by cycling through all of them.

etileb committed:
Update Pioneer-DDJ-400-script.js

Beat FX buttons now change the focused effect, while the FX select button cycles the focused effect.

ronso0 commented on res/controllers/Pioneer-DDJ-400-script.js:
```diff
@@ -44,7 +44,7 @@
 
 var PioneerDDJ400 = {};
-
+engine.setValue("[EffectRack1_EffectUnit1]", "focused_effect", 3);
```
I think this slipped in by accident?
(it's in the wrong place, and also looks like a personal preference since the focused effect is stored in mixxx.cfg)
Please revert.

etileb replied:
No, it was like that before, it makes sure that there is an effect selected on startup.

ronso0 replied:
This line didn't exist, that's way it shows up in the diff. Also, it's in the wrong place.
Please revert.

etileb commented:
> I found the layout confusing with multiple different Buttons for focusing only 3 effects. I changed it so that there is one button for cycling through the effects. I also was annoyed but having to press shift+beat FX to change the focused effect, so now it works without shift.

etileb commented:
> The proposed changes would modify the focus behavior inconsistently, i.e. effect 1 and 2 could still be focused by dedicated buttons while effect 3 could only be reached by cycling through all of them.
> 
> I actually like the cycling apporach, but then the beat < > buttons would need to be remapped.

That's what I actually did.

ronso0 commented:
Please rebase onto mixxx/2.3.
Also, please install and run the pre-commit hooks to resolve the code style issues (see https://github.com/mixxxdj/mixxx/blob/main/CONTRIBUTING.md#git-workflow, 5th bullet point). The change set is small so you can simply squash the fixes and force-push.

etileb committed:
Update Pioneer-DDJ-400-script.js

ywwg commented:
@etileb , if you're unable to do the rebase, I can take over this PR and clean it up for merge.  I don't have a DDJ-400 so I will need someone with that controller to verify that it works

tiesjan commented:
Hi there! Since I own one of these controllers myself and I found the current effects configuration confusing, I've gone ahead and picked up these changes in PR #10912. If someone has the time to have a look, that would be great, thanks! 🙂

daschuer commented:
I think this was replaced by the already merged #10912 and can be closed. Please file a new PR if aspects are missing.
